src/ping.rs

```diff
@@ -43,30 +45,44 @@ pub fn ping(addr: IpAddr, timeout: Option<Duration>, ttl: Option<u32>, ident: Op
 
     socket.set_ttl(ttl.unwrap_or(64))?;
 
-    socket.set_write_timeout(timeout)?;
+    socket.set_write_timeout(Some(timeout))?;
 
     socket.send_to(&mut buffer, &dest.into())?;
 
-    socket.set_read_timeout(timeout)?;
+    // loop until either an echo with correct ident was received or timeout is over
+    let mut time_elapsed = Duration::from_secs(0);
+    loop {
+        socket.set_read_timeout(Some(timeout - time_elapsed))?;
 
-    let mut buffer: [u8; 2048] = [0; 2048];
-    socket.read(&mut buffer)?;
+        let mut buffer: [u8; 2048] = [0; 2048];
+        socket.read(&mut buffer)?;
 
-    let _reply = if dest.is_ipv4() {
-        let ipv4_packet = match IpV4Packet::decode(&buffer) {
-            Ok(packet) => packet,
-            Err(_) => return Err(Error::InternalError.into()),
+        let reply = if dest.is_ipv4() {
+            let ipv4_packet = match IpV4Packet::decode(&buffer) {
+                Ok(packet) => packet,
+                Err(_) => return Err(Error::InternalError.into()),
+            };
+            match EchoReply::decode::<IcmpV4>(ipv4_packet.data) {
+                Ok(reply) => reply,
+                Err(_) => return Err(Error::InternalError.into()),
+            }
+        } else {
+            match EchoReply::decode::<IcmpV6>(&buffer) {
+                Ok(reply) => reply,
+                Err(_) => return Err(Error::InternalError.into()),
+            }
         };
-        match EchoReply::decode::<IcmpV4>(ipv4_packet.data) {
-            Ok(reply) => reply,
-            Err(_) => return Err(Error::InternalError.into()),
-        }
-    } else {
-        match EchoReply::decode::<IcmpV6>(&buffer) {
-            Ok(reply) => reply,
-            Err(_) => return Err(Error::InternalError.into()),
+
+        if reply.ident == request.ident {
+            // received correct ident
+            return Ok(());
         }
-    };
 
-    return Ok(());
+        // if ident is not correct check if timeout is over
+        time_elapsed = SystemTime::now().duration_since(time_start).expect("Clock may have gone backwards");
```

Comment:
Since this is a library, call `expect` is not good here, for it may cause panic for user's process. Return an error will be good.